From d5649ea1d207d6cd2124fe276ff75876540ac758 Mon Sep 17 00:00:00 2001
From: "GONG, Ruiqi" <gongruiqi1@huawei.com>
Date: Thu, 20 Jan 2022 11:07:18 +0800
Subject: [PATCH 1489/2065] [Huawei] selinux: genfscon ns: Add kernel arg
 "selinux_ns_fs"

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186697

----------------------------------------

This kernel argument specifies on which filesystems the genfscon
namespace is enabled, which determines if this feature will be turned
on in runtime.

Signed-off-by: GONG, Ruiqi <gongruiqi1@huawei.com>
---
 security/selinux/hooks.c            | 38 +++++++++++++++++++++++++++++
 security/selinux/include/security.h |  1 +
 2 files changed, 39 insertions(+)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 86159b32921c..9fc4f3161b94 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -152,6 +152,44 @@ static int __init checkreqprot_setup(char *str)
 }
 __setup("checkreqprot=", checkreqprot_setup);
 
+bool selinux_ns_enabled;
+
+enum sel_ns_fs_type {
+	NS_PROCFS = 0,
+	NS_SYSFS,
+	NS_TYPE_NUM,
+};
+
+static bool selinux_ns_fs_enabled[NS_TYPE_NUM];
+
+static int fs_str_idx(const char *str)
+{
+	if (!strcmp(str, "proc"))
+		return NS_PROCFS;
+	else if (!strcmp(str, "sysfs"))
+		return NS_SYSFS;
+	else
+		return NS_TYPE_NUM;
+}
+
+static int __init selinux_ns_fs_setup(char *str)
+{
+	char *token;
+	int idx;
+
+	while ((token = strsep(&str, ","))) {
+		idx = fs_str_idx(token);
+		if (idx != NS_TYPE_NUM) {
+			selinux_ns_fs_enabled[idx] = true;
+			selinux_ns_enabled = true;
+		} else
+			pr_warn("SELinux: Invalid argument. SELinux namespace doesn't support filesystem %s\n",
+				token);
+	}
+	return 1;
+}
+__setup("selinux_ns_fs=", selinux_ns_fs_setup);
+
 /**
  * selinux_secmark_enabled - Check to see if SECMARK is currently enabled
  *
diff --git a/security/selinux/include/security.h b/security/selinux/include/security.h
index 63ca6e79daeb..c4dd4a79add3 100644
--- a/security/selinux/include/security.h
+++ b/security/selinux/include/security.h
@@ -109,6 +109,7 @@ struct selinux_state {
 void selinux_avc_init(struct selinux_avc **avc);
 
 extern struct selinux_state selinux_state;
+extern bool selinux_ns_enabled;
 
 static inline bool selinux_initialized(const struct selinux_state *state)
 {
-- 
2.17.1

